File: sources/perl/03_func_sum.pl
# task 03 func_sum — expected output: 100000000
# build: perl 03_func_sum.pl    run: perl 03_func_sum.pl
# note: the perl interpreter has no inliner and no JIT, so add_one is really called 100000000 times.
use strict;
use warnings;

sub add_one {
    return $_[0] + 1;
}

my $value = 0;
for (my $i = 0; $i < 100000000; $i++) {
    $value = add_one($value);
}

print $value, "\n";

File: sources/ruby/03_func_sum.rb
# task 03 func_sum — expected output: 100000000
# build: none (interpreted)    run: ruby 03_func_sum.rb (cruby) | ruby --yjit 03_func_sum.rb (cruby+yjit) | jruby 03_func_sum.rb (jruby, needs Java 25)
# The stock Windows CRuby build has no YJIT: `ruby --yjit` warns "Ruby was built without YJIT support".
# CRuby's interpreter never inlines add_one; the YJIT/JRuby JITs may inline this call site once it is hot.

def add_one(n)
  n + 1
end

value = 0
i = 0
while i < 100_000_000
  value = add_one(value)
  i += 1
end

puts value
